Expand oracle developer docs (#7290)
ethereum/ethereum-org-website: src/content/developers/docs/oracles/index.md

---
title: Oracles
description: Oracles provide Ethereum smart contracts with access to real-world data, unlocking more use-cases and greater value for users.  
lang: en
sidebar: true
incomplete: true
---

Oracles extract information from the real world and put it on the blockchain for smart contracts to use. This is necessary because smart contracts running on Ethereum cannot access information stored outside the blockchain network. 

> Giving smart contracts the ability to execute based on real-world inputs vastly extends the value of decentralized applications. For instance, decentralized prediction markets rely on oracles to provide information about outcomes with which they can validate predictions from users. If Alice bets 20 ETH on who becomes the next U.S. President, the prediction-market dapp needs an oracle to confirm election results and determine if Alice is eligible for a payout. 

Reviewer: ```suggestion
Giving smart contracts the ability to execute based on real-world inputs extends decentralized applications' value. For instance, decentralized prediction markets rely on oracles to provide information about outcomes with which they can validate user predictions. Suppose Alice bets 20 ETH on who will become the next U.S. President. In that case, the prediction-market dapp needs an oracle to confirm election results and determine if Alice is eligible for a payout.
```

Reviewer: 1. I wonder if the example about Alice is too in the weeds for the introduction paragraphs.
2. The way this section ends feels a little abrupt. Making me feel more like removing it is a good option.

Author: > I wonder if the example about Alice is too in the weeds for the introduction paragraphs.

1. The original page already had this: "For example, prediction market dapps use oracles to settle payments based on events. A prediction market may ask you to bet your ETH on who will become the next president of the United States. They'll use an oracle to confirm the outcome and payout to the winners." The Alice example was to make the concept a bit clearer. 

2. The Alice example doesn't feel too technical in my opinion since we already describe oracles as sourcing information from the real-world for use on-chain. 

> The way this section ends feels a little abrupt. Making me feel more like removing it is a good option.

It does feel *brief*, but I think the intros shouldn't be longer than necessary. This one prepares the reader for the content that follows, so leaving it in might be a good idea.

## Prerequisites {#prerequisites}

This page assumes the reader is familiar with Ethereum fundamentals, including [nodes](/developers/docs/nodes-and-clients/), [consensus mechanisms](/developers/docs/consensus-mechanisms/), and the [EVM](/developers/docs/evm/). You should also have a good grasp of [smart contracts](/developers/docs/smart-contracts/) and [smart contract anatomy](/developers/docs/smart-contracts/anatomy/), especially events. 

## What is a blockchain oracle? {#what-is-a-blockchain-oracle} 

> Oracles are applications that source, verify, and transmit external information (i.e., information stored off-chain) to smart contracts running on the blockchain. Besides “pulling” off-chain data and broadcasting it on Ethereum, oracles can also “push” information from the blockchain to external systems. An example is an oracle that unlocks a smart lock once the user sends the fee via an Ethereum transaction. 

Reviewer: ```suggestion
Oracles are applications that source, verify, and transmit external information (i.e. information stored off-chain) to smart contracts running on the blockchain. Besides “pulling” off-chain data and broadcasting it on Ethereum, oracles can also “push” information from the blockchain to external systems. An example is an oracle that unlocks a smart lock once the user sends the fee via an Ethereum transaction.
```

Reviewer: > An example is an oracle that unlocks a smart lock once the user sends the fee via an Ethereum transaction.

Unclear what this is an example of, it could be:

a) an oracle
b) pulling information
c) pushing information

> Oracles act as a “bridge” connecting smart contracts on blockchains to the real world. Without oracles, smart contract applications would only be able to access on-chain data. A blockchain oracle provides a mechanism for triggering smart contract functions using real-world data.

Reviewer: ```suggestion
Oracles act as a “bridge” connecting smart contracts on blockchains to the real world. Without oracles, smart contract applications would only be able to access on-chain data. An oracle provides a mechanism for triggering smart contract functions using real-world data.
```

> Oracles act as a “bridge” connecting smart contracts and blockchains to the real world. Without oracles, smart contract applications would be limited in usefulness and real-world value. A blockchain oracle provides a mechanism for extracting real-world information which smart contracts can use to trigger state changes in the [Ethereum Virtual Machine](/developers/docs/evm/)

Reviewer 2: ```suggestion
Oracles act as a “bridge” connecting smart contracts and blockchains to the real world. Without oracles, smart contract applications would only be able to access on-chain data. A blockchain oracle provides a mechanism for triggering smart contract functions using real-world data.
```

> Oracles differ based on the source of data (one or multiple sources), trust models (centralized or decentralized) and system architecture (immediate-read, publish-subscribe, and request-response). We can also distinguish between oracles based on whether they retrieve external data for use by on-chain contracts (input oracles), send information from the blockchain to the real world (output orales), or perform computational tasks off-chain (computational oracles).

Reviewer: It feels kinda odd that we are setting up for talking about input, output, and computational oracles here and then don't explicitly comment on these later.

> Oracles differ based on the source of data (one or multiple sources), trust models (centralized or decentralized) and system architecture (immediate-read, publish-subscribe, and request-response). We can also distinguish between oracles based on whether they retrieve external data for use by on-chain contracts (input oracles), send information from the blockchain to the real world (output orales), or perform computational tasks off-chain (computational oracles).

Author: We already talk about computational oracles in some sections, especially ones about generating randomness off-chain, e.g., the following line in the oracle node section: 

> Computational oracles also rely on off-chain nodes to perform intensive computational tasks, which would be impractical to execute on-chain, given gas costs and block size limits. For example, the oracle node may be tasked with generating a verifiably random figure (e.g., for blockchain-based games).  

And input oracles are pretty much everything we cover in most parts (especially w/ price oracles and all). I didn't think it was necessary to go into detail concerning each oracle type since the content implicitly suggests the type of oracle in question at different points (e.g., price oracles are input oracles). Moreover, output oracle use-cases are still few and far in between, so I didn't want to dedicate a lot of time to that part.

> Oracles differ based on the source of data (one or multiple sources), direction of information (inbound or outbound), trust models (centralized or decentralized) and system architecture (immediate-read, publish-subscribe, and request-response). We can also distinguish between oracles that only retrieve external data (input oracles) and ones that perform computation off-chain (computational oracles).

Reviewer 2: I don;t understand the "direction of information" - surely an oracle must have inbound data (real world -> oracle) and outbound data (oracle -> contract) to be useful?

Author: Wanted to distinguish between an oracle that "pushes" information from the blockchain to the real world, e.g., one that triggers a smart lock if the user sends an on-chain transaction and an oracle that "pulls" information to the blockchain (e.g., price oracles). That said, I think "input vs output oracles" is a better way to highlight this detail--will change now,

## Why do smart contracts need oracles? {#why-do-smart-contracts-need-oracles}

> Most developers see smart contracts as simply pieces of code running at specific addresses on the blockchain. However, a more [general view of smart contracts](/smart-contracts/) is that they are self-executing software programs capable of enforcing agreements between parties once specific conditions are met—which explains the term, “smart contracts.”

Reviewer: Nuance: smart contracts automatically execute but don't self-execute. A smart contract won't 'listen' for conditions for it to execute, the information needs to be explicitly pushed to the smart contract.

> But using smart contracts to enforce agreements between people isn't exactly straightforward, given that blockchains are designed to be “deterministic”.  A [deterministic system](https://en.wikipedia.org/wiki/Deterministic_algorithm) is one that always produces the same results given an initial state and a particular input—there is no randomness or variation in the process of computing outputs from inputs. 

Reviewer: We should refer to oracles and smart contracts in relation to Ethereum, instead of blockchains generally. This lets us make assertions (Ethereum is deterministic) instead of reasoning conceptually (blockchains should be deterministic).

```suggestion
But using smart contracts to enforce agreements between people isn't straightforward, given that Ethereum is deterministic. A [deterministic system](https://en.wikipedia.org/wiki/Deterministic_algorithm) is one that always produces the same results given an initial state and a particular input—there is no randomness or variation in the process of computing outputs from inputs.
```

> But there’s a problem with using smart contracts to enforce agreements between people: blockchains are designed to be “deterministic”.  A [deterministic system](https://en.wikipedia.org/wiki/Deterministic_algorithm) is one that always produces the same results given an initial state and a particular input—there is no randomness or variation in the process of computing outputs from inputs. 

Reviewer 2: The start of this paragraph could do with a rethink. It sounds like determinism in the code execution is a bug and oracles are a workaround. That's not the case - oracles are just bridges that enable real world data to influence on-chain events.

Author: Changed to: "But using smart contracts to enforce agreements between people isn't exactly straightforward, given that blockchains are designed to be “deterministic”. "

> To achieve deterministic execution, blockchains limit nodes to reaching consensus on simple binary (true/false) questions using *only* data stored on the blockchain itself. Examples of such questions include: “Did the account owner (identified by a public key) sign this transaction with the paired private key?”, “Does this account have enough funds to cover the transaction?”, “Is this transaction valid in the context of this smart contract?”, and so on. 

Reviewer: To achieve deterministic execution, Ethereum limits nodes to reaching consensus on simple binary (true/false) questions using *only* data stored on the blockchain itself. Examples of such questions include: 

- Did the account owner (identified by a public key) sign this transaction with the paired private key?
- Does this account have enough funds to cover the transaction? 
- Is this transaction valid in the context of this smart contract?

Author: Added this now.

The scenario described earlier also highlights the problem with designing blockchains to pull information from external sources. Oracles, however, solve this problem by taking information from off-chain sources and storing it on the blockchain for smart contracts to consume. Since information stored on-chain is unalterable and publicly available, Ethereum nodes can safely use off-chain data to compute state changes without breaking consensus. 

To do this, an oracle is typically made up of a smart contract running on-chain and some off-chain components. The on-chain contract receives requests for data from other smart contracts, which it passes to the off-chain component (called an oracle node). This oracle node can query data sources—using application programming interfaces (APIs), for example—and send transactions to store the requested data in the smart contract's storage. 

> Essentially, a blockchain oracle bridges the information gap between the blockchain and the external environment, creating “hybrid smart contracts”. A hybrid smart contract is one that functions based on a combination of on-chain contract code and off-chain infrastructure. Decentralized prediction markets, described in the introduction, are an excellent example of hybrid smart contracts.  

Reviewer 2: ```suggestion
Essentially, a blockchain oracle bridges the information gap between the blockchain and the external environment, creating “hybrid smart contracts”. A hybrid smart contract is one that functions based on a combination of on-chain contract code and off-chain infrastructure. Decentralized prediction markets, described in the introduction, are an excellent example of hybrid smart contracts.  Other examples might include crop insurance smart contracts that pay out when a set of oracles determine that certain weather phenomena have taken place.
```

---
title: Oracles
description: Oracles provide Ethereum smart contracts with access to real-world data, unlocking more use-cases and greater value for users.  
lang: en

> sidebar: true

Reviewer: ```suggestion
```

> incomplete: true

Reviewer: ```suggestion
```

---

> Oracles extract information from the real world and put it on the blockchain for smart contracts to use. This is necessary because smart contracts running on Ethereum cannot access information stored outside the blockchain network. 

Reviewer 3: ```suggestion
Oracles are data feeds that bring data from off the blockchain (off-chain) data sources and puts it on the blockchain (on-chain) for smart contracts to use. This is necessary because smart contracts running on Ethereum cannot access information stored outside the blockchain network. 
```

I don't like the use of real-world much if I'm honest. This is all real world, I dont think any data feed is more real world than anything on Ethereum.

> Giving smart contracts the ability to execute based on real-world inputs extends decentralized applications' value. For instance, decentralized prediction markets rely on oracles to provide information about outcomes with which they can validate user predictions. Suppose Alice bets 20 ETH on who will become the next U.S. President. In that case, the prediction-market dapp needs an oracle to confirm election results and determine if Alice is eligible for a payout.

Reviewer 3: ```suggestion
Giving smart contracts the ability to execute using off-chain data inputs extends decentralized applications' value. For instance, decentralized prediction markets rely on oracles to provide information about outcomes with which they can validate user predictions. Suppose Alice bets 20 ETH on who will become the next U.S. President. In that case, the prediction-market dapp needs an oracle to confirm election results and determine if Alice is eligible for a payout.
```

## Prerequisites {#prerequisites}

> This page assumes the reader is familiar with Ethereum fundamentals, including [nodes](/developers/docs/nodes-and-clients/), [consensus mechanisms](/developers/docs/consensus-mechanisms/), and the [EVM](/developers/docs/evm/). You should also have a good grasp of [smart contracts](/developers/docs/smart-contracts/) and [smart contract anatomy](/developers/docs/smart-contracts/anatomy/), especially events. 

Reviewer 3: ```suggestion
This page assumes the reader is familiar with Ethereum fundamentals, including [nodes](/developers/docs/nodes-and-clients/), [consensus mechanisms](/developers/docs/consensus-mechanisms/), and the [EVM](/developers/docs/evm/). You should also have a good grasp of [smart contracts](/developers/docs/smart-contracts/) and [smart contract anatomy](/developers/docs/smart-contracts/anatomy/), especially [events](/glossary/#events). 
```

I think adding a link to the glossary term here would be good.

## What is a blockchain oracle? {#what-is-a-blockchain-oracle} 

> Oracles are applications that source, verify, and transmit external information (i.e. information stored off-chain) to smart contracts running on the blockchain. Besides “pulling” off-chain data and broadcasting it on Ethereum, oracles can also “push” information from the blockchain to external systems. An example of the latter is an oracle that unlocks a smart lock once the user sends the fee via an Ethereum transaction.

Reviewer 3: ```suggestion
Oracles are applications that source, verify, and transmit external information (i.e. information stored off-chain) to smart contracts running on the blockchain. Besides “pulling” off-chain data and broadcasting it on Ethereum, oracles can also “push” information from the blockchain to external systems. An example of the latter could be an oracle that unlocks a smart lock once the user sends the fee via an Ethereum transaction.
```

Changing the wording unless we have an actual example to link out to here.

> Oracles act as a “bridge” connecting smart contracts on blockchains to the real world. Without oracles, smart contract applications would only be able to access on-chain data. An oracle provides a mechanism for triggering smart contract functions using real-world data.

Reviewer 3: ```suggestion
Oracles act as a “bridge” connecting smart contracts on blockchains to off-chain data providers. Without oracles, smart contract applications would only be able to access on-chain data. An oracle provides a mechanism for triggering smart contract functions using off-chain data.
```

> Oracles differ based on the source of data (one or multiple sources), trust models (centralized or decentralized) and system architecture (immediate-read, publish-subscribe, and request-response). We can also distinguish between oracles based on whether they retrieve external data for use by on-chain contracts (input oracles), send information from the blockchain to the real world (output orales), or perform computational tasks off-chain (computational oracles).

Reviewer 3: ```suggestion
Oracles differ based on the source of data (one or multiple sources), trust models (centralized or decentralized), and system architecture (immediate-read, publish-subscribe, and request-response). We can also distinguish between oracles based on whether they retrieve external data for use by on-chain contracts (input oracles), send information from the blockchain to the off-chain applications (output oracles), or perform computational tasks off-chain (computational oracles).
```

}

> ```

Reviewer 2: typos in code snippet (line 125 and 126) lenght --> length

Author: Fixed this now.

## Types of oracles {#types-of-oracles}

### Centralized oracles {#centralized-oracles}

As the name suggests, a centralized oracle is one controlled by a single entity responsible for aggregating off-chain information and updating the oracle contract's data as requested. Centralized oracles are efficient since they rely on a single source of truth. However, using a centralized oracle comes with various problems:

#### Low correctness guarantees {#low-correctness-guarantees}

With centralized oracles, there's no way to confirm if the information provided is correct or not. The oracle provider may be "reputable", but that doesn't eliminate the possibility of someone going rogue or a hacker tampering with the system. If the oracle becomes corrupt, smart contracts will execute based on bad data. 

#### Poor availability {#poor-availability}

Centralized oracles aren't guaranteed to always make off-chain data available to other smart contracts. If the provider decides to turn off the service or a hacker hijacks the oracle's off-chain component, your smart contract is at risk of a denial of service (DoS) attack. 

> ### Poor incentive compatibility {#poor-incentive-compatibility}

Reviewer: ```suggestion
#### Poor incentive compatibility {#poor-incentive-compatibility}
```
